lms/djangoapps/program_enrollments/api/v1/views.py

```diff
@@ -124,3 +133,222 @@ def get(self, request, program_key=None):
+
+        Raises Http404 if the program or course_run does not exist or if the course_run
+        is not in the program
+        """
+        try:
+            course_key = CourseKey.from_string(self.kwargs['course_id'])
+            CourseOverview.get_from_id(course_key)
+        except CourseOverview.DoesNotExist:
+            raise Http404()
+        for course in self.program["courses"]:
+            for course_run in course["course_runs"]:
+                if course_key == CourseKey.from_string(course_run["key"]):
+                    self._course_run = course_run
+                    self._course_key = course_key
+                    return
+        raise Http404()
+
+
+class ProgramCourseEnrollmentsView(ProgramCourseRunSpecificViewMixin, APIView):
+    """
+    A view for enrolling students in a course through a program,
+    modifying program course enrollments, and listing program course
+    enrollments
+
+    Path: /api/v1/programs/{program_uuid}/courses/{course_id}/enrollments
+
+    Accepts: [POST]
+
+    ------------------------------------------------------------------------------------
+    POST
+    ------------------------------------------------------------------------------------
+
+    Returns:
+     * 200: Returns a map of students and their enrollment status.
+     * 207: Not all students enrolled. Returns resulting enrollment status.
+     * 401: User is not authenticated
+     * 403: User lacks read access organization of specified program.
+     * 404: Program does not exist, or course does not exist in program
+     * 422: Invalid request, unable to enroll students.
+    """
+    authentication_classes = (
+        JwtAuthentication,
+        OAuth2AuthenticationAllowInactiveUser,
+        SessionAuthenticationAllowInactiveUser,
+    )
+    permission_classes = (permissions.JWT_RESTRICTED_APPLICATION_OR_USER_ACCESS,)
+    pagination_class = ProgramEnrollmentPagination
+
+    def post(self, request, program_uuid=None, course_id=None):
+        """
+        Enroll a list of students in a course in a program
+        """
+        self.check_course_existence_and_membership()
+        results = {}
+        seen_student_keys = set()
+        enrollments = []
+
+        if not isinstance(request.data, list):
+            raise ValidationError("invalid enrollment record")
+        if len(request.data) > MAX_ENROLLMENT_RECORDS:
+            return Response(
+                'enrollment limit 25', status.HTTP_413_REQUEST_ENTITY_TOO_LARGE
+            )
+
+        try:
+            for enrollment_request in request.data:
+                error_status = self.check_enrollment_request(enrollment_request, seen_student_keys)
+                if error_status:
+                    results[enrollment_request["student_key"]] = error_status
+                else:
+                    enrollments.append(enrollment_request)
+        except (KeyError, ValidationError, TypeError):
```

Comment:
We're catching a whole lot of errors here -- I assume the `KeyError` is in case `"student_key"` isn't in `enrollments`, but that's not clear without a close reading. And what are the `ValidationError` and `TypeError` catches for? 

I recommend doing some refactoring to avoid having to do such a sweeping try/except.

Reply:
yes, key error is if student key isnt there. ValidationError is thrown if the serializer can't parse the data correctly, and TypeError is there to cover the case that we have an input like 
```
[
   {<correctly formatted enrollment request>},
   {<correctly formatted enrollment request>},
   [<whoops this is a list accidentally rather than a dict>],
   or 
   "whoops this is just text",
]
```
rather than throwing the key error when we try to do `student_key = enrollment['student_key']` it'll throw a TypeError

Should I maybe be returning different responses for these different types of errors?

Reply:
Gotcha. Thanks for enumerating those for me. I think the 422 responses are fine. It looks like your code will work correctly here, but I'm wondering if there's a way to write it that'd be easier for a future developer to parse.

Maybe this could be refactored by doing something like:
```python
enrollments_serializer = ProgramCourseEnrollmentRequestSerializer(data=enrollment, many=True)
# enrollments_serializer.errors is now a list of dicts describing the validation errors in each item
# if enrollments_serializer.errors is None, you know you won't get any KeyErrors or TypeErrors
if enrollments_serializer.errors:
   ... # Handle validation errors

   
```
Of course, you'd still have to check for duplicates and conflicts.

If you find that this doesn't make the code any cleaner, then stick with what you have but please briefly note in a comment the source of each of the errors you're catching there.

Reply:
Unfortunately, this doesn't really prevent KeyErrors, since we're checking for invalid statuses in the serialize, so not all errors are instant-fails. I could avoid KeyErrors by checking for the key, but I believe this is the more "pythonic" way.

I think I'll break up the exceptions so that I can comment each one individually, I think that will make it a bit more readable